File: DataStructure/BinaryTree/106_in_post_buildTree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def helper(in_left, in_right):
            if in_left > in_right:
                return None
            
            val = postorder.pop()
            idx = idx_map[val]
            
            tree = TreeNode(val)
            # 右子树 左子树
            tree.right = helper(idx + 1, in_right)
            tree.left = helper(in_left, idx - 1)
            return tree

        # 前后名称要一样
        idx_map = {val:idx for idx, val in enumerate(inorder)}
        return helper(0, len(inorder) - 1)

File: DataStructure/BinaryTree/106_in_post_buildTree.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* helper(const vector<int>& inorder, const vector<int>& postorder, int in_st, int in_end, int post_st, int post_end) {
        if (post_st > post_end)
            return nullptr;
        // if (in_st > in_end)
        //   return nullptr; 都是可以的
        TreeNode *root = new TreeNode(postorder[post_end]);
        int idx = in_st;
        for (int i = in_st; i <= in_end; i++) {
            if (postorder[post_end] == inorder[i]) {
                idx = i;
                break;
            }
        }

        int leftlength = idx - in_st;
        root->left = helper(inorder, postorder, in_st, idx - 1, post_st, post_st + leftlength - 1);
        root->right = helper(inorder, postorder, idx + 1, in_end, post_st + leftlength, post_end - 1);
        return root;
    }

    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return helper(inorder, postorder, 0, inorder.size() - 1, 0, postorder.size() - 1);
    }
};

int main() {
    Solution s;
    // vector<int> pre = {3, 9, 20, 15, 7};
    vector<int> in = {9, 3, 15, 20, 7};
    vector<int> post = {9, 15, 7, 20, 3};
    s.buildTree(in, post);
}